### PR TITLE
[SPARK-32796][SQL] Make withField API support nested struct in array

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch adds nested struct support to `Column.withField` API.

### Why are the changes needed?

Currently `Column.withField` only supports `StructType`. For nested struct in `ArrayType`, it doesn't support. We can support nested struct in array to make the API more general and useful.

### Does this PR introduce _any_ user-facing change?

Yes. Adding nested struct support to `Column.withField` API.

### How was this patch tested?

Unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveWithFields.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Resolves `UnresolvedWithFields`.
+ */
+object ResolveWithFields extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+    case e if !e.childrenResolved => e
+
+    case q: LogicalPlan =>
+      q.transformExpressions {
+        case expr if !expr.childrenResolved => expr
+        case e: UnresolvedWithFields => WithFields(e.col, e.fieldName, e.expr)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveWithFields.scala:34]: This can be moved to other proper rule.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 2]
```diff
@@ -1537,4 +1576,230 @@ class ColumnExpressionSuite extends QueryTest with SharedSparkSession {
       Row(3) :: Row(null):: Nil,
       StructType(Seq(StructField("a", IntegerType, nullable = true))))
   }
+
+  test("withField should add field to struct of array") {
+    checkAnswerAndSchema(
+      arrayLevel1.withColumn("a", 'a.withField("d", lit(4))),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala:1582]: @viirya, BTW, is it same as:

```scala
transform($"struct_col", _.withField("new_col", col))
```
?

**Comment by HyukjinKwon** [comment]
I personally more prefer an explicit call such as `transform($"struct_col", _.withField("new_col", col))` more as it's more explicit but I got that this can be sort of similar with  the nested column access in SQL like `a.b` supports the arrays of structs. No string opinion on this. WDYT @cloud-fan.

**Comment by cloud-fan** [comment]
I agree with @HyukjinKwon that it's better to be more explicit. `a.b` is a bad example, as it arbitrarily decides to support only one level, array of array of struct is not supported. I prefer to not make the same mistake in `withField`.

**Comment by viirya** [comment]
An example looks like:

```scala
private lazy val arrayType = ArrayType(
    StructType(Seq(
      StructField("a", IntegerType, nullable = false),
      StructField("b", IntegerType, nullable = true),
      StructField("c", IntegerType, nullable = false))),
    containsNull = true)

private lazy val arrayStructArrayLevel1: DataFrame = spark.createDataFrame(
    sparkContext.parallelize(Row(Array(Row(Array(Row(1, null, 3)), null, 3))) :: Nil),
    StructType(
      Seq(StructField("a", ArrayType(
        StructType(Seq(
          StructField("a", arrayType, nullable = false),
          StructField("b", IntegerType, nullable = true),
          StructField("c", IntegerType, nullable = false))),
        containsNull = false)))))
```

The data looks like:
```
+---------------------------+
|a                          |
+---------------------------+
|[{[{1, null, 3}], null, 3}]|
+---------------------------+
```

In order to replace deeply nested `b` column, like:
```
+------------------------+
|a                       |
+------------------------+
|[{[{1, 2, 3}], null, 3}]|
+------------------------+
```

Currently by using `transform` +  `withField`, we probably need:

```scala
arrayStructArrayLevel1.withColumn("a",
      transform($"a", _.withField("a",
        flatten(transform($"a.a", transform(_, _.withField("b", lit(2))))))))
```

Using modified `withField`, we can do it like:
```scala
arrayStructArrayLevel1.withColumn("a", $"a".withField("a.b", lit(2)))
```

**Comment by cloud-fan** [comment]
We can save more code by supporting array of array of struct. It's a trade-off between "clear and simple semantic" vs "flexiblity of supporting various input types".

**Comment by viirya** [comment]
@cloud-fan Sorry if I mis-read your comment. Do you mean we should support array of array of struct?

**Comment by cloud-fan** [comment]
I mean we should prefer "clear and simple semantic", otherwise people can always ask to be more flexible and save more code, like supporting array of array of struct.

**Comment by viirya** [comment]
Okay, I see. It also makes sense to me. This is a hard trade-off between simplicity and flexibility. I will close this now. If we need this flexibility in the future, we can revisit it.

**Comment by cloud-fan** [comment]
Does `transform($"struct_col", _.withField("new_col", col))` work for you?

**Comment by encarvlucas** [comment]
Unfortunately it doesn't. I'm using spark through the pyspark library

**Comment by cloud-fan** [comment]
@HyukjinKwon shall we add the `transform` API in pyspark?

**Comment by encarvlucas** [comment]
There is a `transform` implementation in pyspark, but it does not work with python UDFs. Might be related to this [issue](https://issues.apache.org/jira/browse/SPARK-27052).
Also, there is no `withField` function in Spark SQL.

**Comment by cloud-fan** [comment]
Do you mean the `col` in `transform($"struct_col", _.withField("new_col", col))` is a python udf? I'm not sure if this PR can help due to the limitation of python udf execution...

**Comment by encarvlucas** [comment]
Yes. Nevermind then, I retract my request.

**Pull request closed without merging**